src/modules/roc_netio/target_uv/roc_netio/udp_receiver_port.cpp

```diff
@@ -91,65 +96,57 @@ bool UDPReceiver::start(packet::Address& bind_address) {
     }
 
     roc_log(LogInfo, "udp receiver: opened port %s",
-            packet::address_to_str(bind_address).c_str());
+            packet::address_to_str(address_).c_str());
+
+    recv_started_ = true;
 
-    address_ = bind_address;
     return true;
 }
 
-void UDPReceiver::stop() {
-    if (!handle_initialized_) {
-        return;
+void UDPReceiverPort::async_close() {
+    if (closed_) {
+        return; // handle_closed() was already called
     }
 
-    if (uv_is_closing((uv_handle_t*)&handle_)) {
+    if (!handle_initialized_) {
+        closed_ = true;
+        close_handler_.handle_closed(*this);
+
         return;
     }
 
     roc_log(LogInfo, "udp receiver: closing port %s",
             packet::address_to_str(address_).c_str());
 
-    if (int err = uv_udp_recv_stop(&handle_)) {
-        roc_log(LogError, "udp receiver: uv_udp_recv_stop(): [%s] %s", uv_err_name(err),
-                uv_strerror(err));
+    if (recv_started_) {
+        if (int err = uv_udp_recv_stop(&handle_)) {
+            roc_log(LogError, "udp receiver: uv_udp_recv_stop(): [%s] %s",
+                    uv_err_name(err), uv_strerror(err));
+        }
     }
 
     uv_close((uv_handle_t*)&handle_, close_cb_);
```

Comment:
I suggest to guard this with `if (!uv_is_closing(...))`. Just in case. The flow is quite complicated so I would be twice as careful here. Also, I would add `recv_started_ = false` to the above `if` which calls `uv_udp_recv_stop`.

With this two changes calling UDPReceiverPort::async_close() twice will not be harmful (even it's not intended to do so).

BTW note that UDPSenderPort::async_close() *has* such guards.

Reply:
Done